### PR TITLE
Extract bispectrum gradients in compute_snap without summing over neighbors.

### PR DESCRIPTION
**Pull request by rohskopf**
**Summary**

This PR adds a `compute snap` option to extract individual bispectrum gradients for each atom, rather than summing the gradients over neighbors as done currently. This is achieved with a new `dgradflag` keyword to `compute snap`. This feature is necessary when training neural network potentials with bispectrum components.

**Related Issue(s)**

None

**Author(s)**

Drew Rohskopf (SNL), Aidan Thompson (SNL)

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

This change adds an optional keyword that does not break backward compatibility. 

**Implementation Notes**

This PR makes use of the current machinery in `compute_snap.cpp` to calculate bispectrum gradients. We add an option that changes how these gradients are processed. The current `compute_snap.cpp` sums the gradients over neighbors, and here we do not perform that sum if `dgradflag=1`. This option also changes the structure of the global snap array completely, as explained in the new docs. 

Correctness is verified with an example PyLammps python script in `examples/snap/compute_snap_dgrad.py`. This example compares the sum of bispectrum gradients using the original/default method (`dgradflag=0`) and the sum of bispectrum components using our new method (`dgradflag=1`). For `dgradflag=1`, this python script manually sums the bispectrum gradients over neighbors and compares with the gradient sums calculated in LAMMPS for `dgradflag=0`. This python script also compares the reference forces and energies calculated from these two options, verifying that the new `dgradflag=1` option contains the same data as the default/traditional `dgradflag=0` option. 

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [x] One or more example input decks are included

### CODE REVIEW

File: src/ML-SNAP/compute_snap.cpp [hunk 1]
```diff
@@ -142,9 +142,15 @@ ComputeSnap::ComputeSnap(LAMMPS *lmp, int narg, char **arg) :
       if (iarg + 2 > narg) error->all(FLERR, "Illegal compute {} command", style);
       bikflag = utils::inumeric(FLERR, arg[iarg + 1], false, lmp);
       iarg += 2;
-    } else if (strcmp(arg[iarg], "switchinnerflag") == 0) {
-      if (iarg + 2 > narg) error->all(FLERR, "Illegal compute {} command", style);
-      switchinnerflag = utils::inumeric(FLERR, arg[iarg + 1], false, lmp);
+    } else if (strcmp(arg[iarg],"dgradflag") == 0) {
+      if (iarg+2 > narg)
+        error->all(FLERR,"Illegal compute snap command");
+      dgradflag = atoi(arg[iarg+1]);
```
Comment (akohlmey) [inline src/ML-SNAP/compute_snap.cpp]: This should use `utils::inumeric()` and not `atoi()`. Same below.

File: src/ML-SNAP/compute_snap.cpp [hunk 2]
```diff
@@ -13,7 +13,6 @@
 ------------------------------------------------------------------------- */
 
 #include "compute_snap.h"
-
```
Comment (akohlmey) [inline src/ML-SNAP/compute_snap.cpp:16]: Please don't delete this line. It is part of the include file conventions and required once we enable clang-format globally, so that it doesn't sort the class header away from the top spot.

File: src/ML-SNAP/compute_snap.cpp [hunk 3]
```diff
@@ -208,20 +228,21 @@ ComputeSnap::ComputeSnap(LAMMPS *lmp, int narg, char **arg) :
 
 ComputeSnap::~ComputeSnap()
 {
+
```
Comment (akohlmey) [inline src/ML-SNAP/compute_snap.cpp]: Please undo the whitespace changes here. We want to keep the programming style as consistent as possible throughout LAMMPS to make it easier for people to read and modify. That sometimes requires to against personal preferences.

File: src/ML-SNAP/compute_snap.cpp [hunk 4]
```diff
@@ -231,8 +252,9 @@ void ComputeSnap::init()
   if (force->pair == nullptr)
     error->all(FLERR,"Compute snap requires a pair style be defined");
 
-  if (cutmax > force->pair->cutforce)
+  if (cutmax > force->pair->cutforce){
```
Comment (akohlmey) [inline src/ML-SNAP/compute_snap.cpp]: See my comment above about programming style. That includes not adding curly braces for a single statement. If you do add them, then there should be a blank before the curly brace. The following page summarizes the most important conventions:
https://docs.lammps.org/Modify_style.html#programming-style-requirements-varied

File: src/ML-SNAP/compute_snap.cpp [hunk 5]
```diff
@@ -298,12 +321,13 @@ void ComputeSnap::compute_array()
...
-  for (int irow = 0; irow < size_array_rows; irow++)
-    for (int icoeff = 0; icoeff < size_array_cols; icoeff++)
+  for (int irow = 0; irow < size_array_rows; irow++){
+    for (int icoeff = 0; icoeff < size_array_cols; icoeff++){
       snap[irow][icoeff] = 0.0;
+    }
+  }
 
   // clear local peratom array
-
```
Comment (akohlmey) [inline src/ML-SNAP/compute_snap.cpp:306]: @sjplimp specifically dislikes removing blank lines after a comment.

File: src/ML-SNAP/compute_snap.cpp [hunk 6]
```diff
@@ -298,12 +321,13 @@ void ComputeSnap::compute_array()
 
   // clear global array
 
-  for (int irow = 0; irow < size_array_rows; irow++)
-    for (int icoeff = 0; icoeff < size_array_cols; icoeff++)
+  for (int irow = 0; irow < size_array_rows; irow++){
+    for (int icoeff = 0; icoeff < size_array_cols; icoeff++){
       snap[irow][icoeff] = 0.0;
+    }
+  }
```
Comment (akohlmey) [inline src/ML-SNAP/compute_snap.cpp]: See above

**Review by akohlmey — CHANGES_REQUESTED** [review]
Please see my comments. Those are all cosmetic changes to better align the programming style to the general LAMMPS conventions.

**Comment by rohskopf** [comment]
Thanks @akohlmey. Made some cosmetic changes.

**Review by athomps — APPROVED** [review]
I approve.

**Review by akohlmey — APPROVED** [review]
